Test and fix the notebook tester

ceball opened the pull request:
# Significant behavior changes

* [x] Fix missing raise in "vendored tests" (from nbval). Fixes #91.
* [x] Ensure cells are not executed multiple times in the same kernel. Fixes #94.

# tests_vendored cleanup

* [x] ~~Support always starting a new kernel per test (with option for previous behavior).~~ Always start a new kernel per test. Future work: optimize by tracking cell execution so that each cell is only executed once in any given context. #105 
* [x] Fix tests_vendored method names and make class methods actually be class methods

# New tests

* [x] ~~Test cumulative cell run state (shared kernel)~~ (we are dropping this current behavior)
* [x] Test cumulative cell run state (fresh kernel)
* [x] Test with new process per test
* [x] Test cell exception
* [x] Test test exception
* [x] Test test failure

# Cleanup internal to this PR

* [x] Clean up test_test.py (e.g. fix the enterprise class names, make it possible to understand what is being done, generate the test notebooks, etc - see TODOs).
* [x] ~~If keeping optional 'multiple executions of a cell in same kernel' behavior, make option easily selectable.~~ we are not keeping this behavior

ceball commented on nbcelltests/lint.py:
```diff
@@ -158,16 +158,17 @@ def run(notebook, executable=None, rules=None):
         exp = ScriptExporter()
         (body, resources) = exp.from_notebook_node(nb)
+        tf = NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
+        tf_name = tf.name
```
changes in this file are unrelated to this PR - will move them

timkpaine commented:
@ceball did you want to upgrade this from draft PR to real PR?

ceball commented:
@timkpaine before I finish this PR, please tell me your opinion about #94. I.e. is it ok to modify the existing behavior so that each test is now run in a fresh kernel (rather than the current behavior of running all tests in the same kernel*)?

* unless tests are all distributed to independent processes (behavior in this case unchanged: fresh kernel per test)

timkpaine commented:
@ceball I think its fine, maybe @vidartf has an opinion

ceball commented:
@timkpaine @vidartf I will proceed with that change, but Vidar and I will review the behavior before the next release so we can revert if necessary. Hope that's ok.

ceball commented:
I'd prefer a squash merge here (no need to preserve my false accusations about azure pipelines 😂). Actually I prefer squash merges in general as an easy way to have a nice history on master, but I'm not precious about it and understand the drawbacks of squash merge.

ceball commented:
@vidartf and I have planned out how we'd like to change the behavior of celltests. The planned changes are an optimization on top of this PR, so I have now completely dropped the ability to go back to the existing behavior (i.e. the behavior before this PR). Apart from dropping ability to select old behavior by option, there have been no significant changes to this PR since my most recent comment 3 days ago.

vidartf commented on nbcelltests/tests/test_test.py:
```diff
@@ -5,9 +5,223 @@
+    spec = importlib.util.spec_from_file_location(pypath, f)
+    mod = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+    return mod
```
While I appreciate that this is just our testing code, given my previous experiences with importlib, I would strongly prefer this to be tested on the full 3.5 - 3.8 set of Python versions. Not a blocker though, just a heads up that there might be some differences across versions.

ceball replied:
Ah, there was a TODO comment to exactly this effect. I wonder where it went. I'll restore it!

ceball replied:
> that this is just our testing code

What kind of attitude is that in a library about testing code, ha ha 😲

ceball replied:
It's like how I wanted to turn off the linter while I was working on the linting part.

vidartf commented on nbcelltests/tests/test_test.py:
```diff
@@ -5,9 +5,223 @@
+        tf = tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
+        tf_name = tf.name
+        try:
+            cls.generated_tests = _import_from_path("nbcelltests.tests.%s.%s" % (__name__, cls.__name__), run(cls.NBNAME, filename=tf_name))
```
This is a pretty obtuse line of code. Some comments would be nice :) E.g.:
- Why is the class name included in the import location spec? Is this merely for display purposes, or does it require class names to follow some formula?
- Maybe write a little about what `generated_tests` will be after this is called, and how it should be used.

ceball replied:
Agree with your first bullet point (and that the line is obtuse).

For the second bullet point, is the existing method docstring enough?

vidartf commented on nbcelltests/tests/test_test.py:
```diff
@@ -5,9 +5,223 @@
+        0:           -       -       -
+
+        1: either   x=0      -      x=0   (expected)
+               or    -       -      x>0   (bad; x was previously defined in kernel)
```
What would trigger the `or` condition of this either/or branching? Can you add a short comment on the top describing how to read the table?

ceball replied:
It's tricky - if you can suggest something better I'll be grateful. 

If `x` is already defined, nothing is done. If `x` is not defined, it's set to 0. This is how I caught the case that the tests were running all the cells multiple times in the same kernel.

vidartf replied:
> If x is already defined, nothing is done. If x is not defined, it's set to 0. 

How do you read this out of the table though? 😕

ceball replied:
I said it was tricky! :) I'll think about it - I'm sure we can do better.

vidartf commented on nbcelltests/tests_vendored.py:
```diff
@@ -46,10 +46,23 @@
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.kernel.stop()
```
What is the class-level kernel used for?

ceball replied:
We ultimately want one kernel per class (i.e. one kernel per set of tests being run).

The hack we currently have to ensure not executing a cell multiple times in the same kernel is:

```
    # TODO: starting a new kernel per test is expensive, and
    # could be optimized.
    def setUp(self):
        self.kernel = RunningKernel(self.KERNEL_NAME)
```

So, starting a kernel in the `setUpClass` method is the one that will remain - starting a kernel in `setUp` will be removed.

Does that make sense? Or I am confused...?

ceball replied:
(Of course, you're right that at the moment the class level kernel is pointless, as it's immediately replaced by each test. But I'd rather keep the class level kernel for now, as it's what should be there if we weren't hacking things with a new kernel per test.)

vidartf replied:
Ok, so we keep it in this diff as a forward looking change :)

vidartf commented on setup.py:
```diff
@@ -29,7 +29,8 @@
     'mock',
-    'autopep8'
+    'autopep8',
+    'pytest-xdist ; sys_platform != "win32"'
```
Is it really needed to use the `--forked` option (i.e. vs. just `-n2`)? If we could just use `-n2`, then the distributed test would also work on Windows. The state of the `--forked` option seems confused as well, as it has been spun out of xdist into pytest-forked, but that is listed as unmaintained.

ceball replied:
Originally, before I made changes to the code, and had only written tests, I used `--forked` to run the new tests to verify that the new tests passed when there was **one test per process** (because that would guarantee one kernel per test). `--forked` is the way I could see to do that easily. Now I have made changes to the code, `--forked` is not really necessary at all any more and could actually be removed.

After the subsequent optimization is applied, we will be more interested in just distributing the tests in some way like you say with half going somewhere and half going somewhere else. So would you accept your proposed change during that subsequent optimization work instead?

ceball replied:
https://github.com/jpmorganchase/nbcelltests/issues/107

ceball commented:
@vidartf I believe I've addressed the points you raised; please let me know if I missed anything, or otherwise please explicitly approve (a requirement for merging).

vidartf commented on nbcelltests/tests/test_test.py:
```diff
@@ -68,21 +68,29 @@ class TestCumulativeRun(_TestCellTests):
+        2:    -       -      x>0   (bad; x was previously defined in kernel)
+        3:   x+=1     -      x>1   (bad)
+        4:   x+=1     -      x>2   (bad)
+        5:    -      x+=1    x>3   (bad)
```
Thanks this is much better :)
